server/src/main/java/org/apache/iotdb/db/rescon/SystemInfo.java

```diff
@@ -0,0 +1,247 @@
+
+  /**
+   * Report applying a new out of buffered array to system. Attention: It should be invoked before
+   * applying new OOB array actually.
+   *
+   * @param dataType data type of array
+   * @param size     size of array
+   * @return Return true if it's agreed when memory is enough.
+   */
+  public synchronized boolean applyNewOOBArray(TSDataType dataType, int size) {
+    // if current memory is enough
+    if (arrayPoolMemCost + totalSgMemCost + dataType.getDataTypeSize() * size
+        < config.getAllocateMemoryForWrite() * FLUSH_PROPORTION) {
+      arrayPoolMemCost += dataType.getDataTypeSize() * size;
+      logger.debug("Current total mem cost is {}", (arrayPoolMemCost + totalSgMemCost));
+      rejected = false;
+      return true;
+    } else if (arrayPoolMemCost + totalSgMemCost + dataType.getDataTypeSize() * size
+        < config.getAllocateMemoryForWrite() * REJECT_PROPORTION) {
+      arrayPoolMemCost += dataType.getDataTypeSize() * size;
+      // invoke flush()
+      logger.debug("Out of buffer arraies are too large, call for flushing. "
+          + "Current total mem cost is {}", (arrayPoolMemCost + totalSgMemCost));
+      rejected = false;
+      flush();
+      return true;
+    } else {
+      logger.debug("Out of buffer arraies are too large, call for flushing "
+          + "and change system to rejected status...Current total mem cost is {}", 
+          (arrayPoolMemCost + totalSgMemCost));
+      rejected = true;
+      flush();
+      return false;
+    }
+  }
+
+  /**
+   * Report current mem cost of storage group to system.
+   *
+   * @param storageGroupInfo storage group
+   * @param delta mem cost
+   */
+  public synchronized void reportStorageGroupStatus(StorageGroupInfo storageGroupInfo) {
+    long delta = storageGroupInfo.getSgMemCost() - 
+        reportedSgMemCostMap.getOrDefault(storageGroupInfo, 0L);
+    totalSgMemCost += delta;
+    logger.debug("Report Storage Group Status to system. "
+        + "Current array pool mem cost is {}, sg mem cost is {}.", arrayPoolMemCost,
+        totalSgMemCost);
+    reportedSgMemCostMap.put(storageGroupInfo, storageGroupInfo.getSgMemCost());
+
+    long newSgReportThreshold = calculateNewSgReportThreshold(storageGroupInfo);
+    storageGroupInfo.setStorageGroupReportThreshold(newSgReportThreshold);
+    if (arrayPoolMemCost + totalSgMemCost
+        >= config.getAllocateMemoryForWrite() * FLUSH_PROPORTION) {
+      logger.debug("The total storage group mem costs are too large, call for flushing. "
+          + "Current sg cost is {}, array pool cost is {}", totalSgMemCost, arrayPoolMemCost);
+      flush();
+    } 
+    if (arrayPoolMemCost + totalSgMemCost
+        >= config.getAllocateMemoryForWrite() * REJECT_PROPORTION) {
+      logger.debug("Change system to reject status...");
+      rejected = true;
+    }
+  }
+
+  /**
+   * Calculate new Sg report threshold. The maximum value is 
```

Comment:
the maximum value is what?

Reply:
This part is what I'm fixing. 
Reporting sg to systemInfo is only invoked when insert Text data. It's necessary to set a limit to sgReportThreshold, otherwise the sgReportThreshold will be too large and cause OOM. 
I'm trying to find a better way to limit the sgReportThreshold.